Document Leak occurs in LayoutTests/editing/mac/selection/word-thai.html

NathanSolomon733 opened the pull request:
#### 72e418d5143c735ed4c0f2f89b0ee5c64c15a5e0
<pre>
Document Leak occurs in LayoutTests/editing/mac/selection/word-thai.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=280688">https://bugs.webkit.org/show_bug.cgi?id=280688</a>
<a href="https://rdar.apple.com/137058331">rdar://137058331</a>

Reviewed by Ryan Reno.

The document leak occurs because the m_dragStartSelection member in
EventHandler, which is of type SimpleRange, holds Refs to Nodes.
This is a problem because the Document holds Refs to these nodes
because the document holds a strong reference to an Editor, which holds a
strong reference to a LocalFrame, which holds a strong reference to an
EventHandler.
This causes a reference cycle to occur between
the Document and the Nodes the EventHandler holds. Fix this reference
cycle by changing the type of m_dragStartSelection to a new type
that allows the member to hold a WeakReference to these nodes. Creating
this new type instead of modifying SimpleRange to hold WeakPtrs to nodes
prevents making extensive changes to areas of the codebase that can
still use a SimpleRange without causing a
reference cycle. Keep the change isolated to what code
relies on accessing m_dragStartSelection&apos;s Nodes to prevent adding
extensive amounts of null checking in other areas.

The test I ran to verify the fix was: run-webkit-tests --world-leaks LayoutTests/editing/mac/selection/word-thai.html
This test would previously fail when --world-leaks was passed in.

* Source/WebCore/dom/BoundaryPoint.h:
(WebCore::BoundaryPointWeak::BoundaryPointWeak):
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::SimpleRangeWeak::SimpleRangeWeak):
(WebCore::firstIntersectingNode):
(WebCore::nodePastLastIntersectingNode):
(WebCore::firstIntersectingNodeWithDeprecatedZeroOffsetStartQuirk):
(WebCore::commonInclusiveAncestor):
(WebCore::IntersectingNodeIterator::IntersectingNodeIterator):
* Source/WebCore/dom/SimpleRange.h:
(WebCore::SimpleRangeWeak::startContainer const):
(WebCore::SimpleRangeWeak::protectedStartContainer const):
(WebCore::SimpleRangeWeak::startOffset const):
(WebCore::SimpleRangeWeak::endContainer const):
(WebCore::SimpleRangeWeak::protectedEndContainer const):
(WebCore::SimpleRangeWeak::endOffset const):
(WebCore::SimpleRangeWeak::collapsed const):
(WebCore::SimpleRange::operator SimpleRangeWeak const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateSelectionForMouseDrag):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::treeOrderBoundaryPoints):

Canonical link: <a href="https://commits.webkit.org/285629@main">https://commits.webkit.org/285629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75033bd8774c867a82eb61707fd48ffe69d90424

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72982 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52411 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25790 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77185 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24220 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75097 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60216 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57366 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15854 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47377 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37784 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44021 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20284 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22549 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65879 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20640 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78859 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65819 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65089 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16144 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7075 "Passed tests") | | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2998 "Built successfully") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

NathanSolomon733 commented on Source/WebCore/dom/SimpleRange.cpp:
```diff
@@ -102,6 +114,62 @@ IntersectingNodeIterator::IntersectingNodeIterator(const SimpleRange& range)
+{
+    enforceEndInvariant();
+}
+
```
For SimpleRangeWeak I overloaded these functions that were provided by SimpleRange. The logic is identical except that it checks the passed in SimpleRangeWeak if the nodes the SimpleRangeWeak points to is non-null. They are not currently being used though by whoever needs to access the members of m_dragStartSelection. Should these just not be included in the PR because they are not currently needed? Or, should I define all the equivalent methods that SimpleRange implements to operate on SimpleRanges so the two can have similar usages if ever needed in the future?

NathanSolomon733 commented on Source/WebCore/dom/SimpleRange.h:
```diff
@@ -29,6 +29,47 @@
+    SimpleRangeWeak(const BoundaryPointWeak&, const BoundaryPointWeak&);
+    SimpleRangeWeak(BoundaryPointWeak&&, BoundaryPointWeak&&);
+    SimpleRangeWeak(const BoundaryPoint&&, const BoundaryPoint&&);
+};
```
For a struct like this, do we typically make a new .h and .cpp file for it? It is similar to a SimpleRange which is why I put it in SimpleRange.h, but I wasn't sure if there is a policy to follow when to make dedicated header and implementation files.

NathanSolomon733 commented on Source/WebCore/dom/SimpleRange.h:
```diff
@@ -29,6 +29,47 @@
+
+    bool collapsed() const { return start == end; }
+
+    friend bool operator==(const SimpleRangeWeak&, const SimpleRangeWeak&) = default;
```
These methods are not actually used by the only instance of SimpleRangeWeak which is in m_dragStartSelection. SimpleRange had these functions implemented, and I was unsure if I should just fill out this functionality now or leave it for the future, where these methods would be implemented on an as-needed basis.

rreno replied:
I think we should not have them if they're unused. They're also kind of awkward. An `optional<Ref>` behaves functionally like a `RefPtr` in that either it has a value or it doesn't. Same with `optional<std::reference_wrapper<Node>>` that's just a `Node*`.

If we don't need the functions then I don't think we should include them in this patch.

NathanSolomon733 replied:
Done.

NathanSolomon733 commented on Source/WebCore/dom/SimpleRange.h:
```diff
@@ -42,6 +83,10 @@ struct SimpleRange {
+    operator SimpleRangeWeak() const 
+    {
+        return { BoundaryPointWeak(start.container.get(), start.offset), BoundaryPointWeak(end.container.get(), end.offset) };
+    }    
```
This operator we actually need. It's used on line 552 of EventHandler.cpp to perform an implicit conversion from SimpleRange to SimpleRangeWeak.

rreno replied:
I think we should have an explicit conversion function instead. Something like `makeWeakSimpleRange`.

NathanSolomon733 replied:
Done.

rniwa commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -672,7 +672,7 @@ class EventHandler final : public CanMakeCheckedPtr<EventHandler> {
 #if ENABLE(DRAG_SUPPORT)
     LayoutPoint m_dragStartPosition;
-    std::optional<SimpleRange> m_dragStartSelection;
+    std::optional<SimpleRangeWeak> m_dragStartSelection;
```
I'd call this WeakSimpleRange instead.

NathanSolomon733 replied:
Done.

rniwa commented on Source/WebCore/dom/BoundaryPoint.h:
```diff
@@ -79,4 +79,22 @@ inline BoundaryPoint makeBoundaryPointAfterNodeContents(Node& node)
     return { node, node.length() };
 }
 
+struct BoundaryPointWeak {
```
I'd call this WeakBoundaryPoint instead.

rniwa commented on Source/WebCore/dom/SimpleRange.cpp:
```diff
@@ -102,6 +114,62 @@ IntersectingNodeIterator::IntersectingNodeIterator(const SimpleRange& range)
+    enforceEndInvariant();
+}
+
+IntersectingNodeIterator::IntersectingNodeIterator(const SimpleRangeWeak& range, QuirkFlag)
```
Is this actually used somewhere?

NathanSolomon733 replied:
Oh no it isn't. Those two constructors that take in SimpleRangeWeak can be removed.

rniwa replied:
Let's remove them then.

NathanSolomon733 replied:
Done.

rreno reviewed (commented):
I'd include the test you ran manually to verify the fix in the commit message. Something like `run-webkit-tests --world-leaks editing/mac/selection/word-thai.html` I think?

rreno commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1095,8 +1095,8 @@ void EventHandler::updateSelectionForMouseDrag(const HitTestResult& hitTestResul
-        else if (newSelection.isBaseFirst() && !oldSelection.isBaseFirst() && oldSelection.start() < newSelection.start() && m_dragStartSelection) {
-            VisibleSelection dragStartSelection { *m_dragStartSelection };
+        else if (newSelection.isBaseFirst() && !oldSelection.isBaseFirst() && oldSelection.start() < newSelection.start() && m_dragStartSelection && m_dragStartSelection.value().start.container && m_dragStartSelection.value().end.container) {
+            VisibleSelection dragStartSelection { { BoundaryPoint(*(m_dragStartSelection.value().start.container), m_dragStartSelection.value().start.offset), BoundaryPoint(*(m_dragStartSelection.value().end.container), m_dragStartSelection.value().end.offset) } };
```
This is hard to read. Could this be factored into a lambda or a private member function that computes all this stuff and has a descriptive name?

NathanSolomon733 replied:
If you were referring to just the line constructing the SimpleRange to pass into the VisibleSelection constructor, yes. I defined a private member in EventHandler called createSimpleRangeFromDragStartSelection().

rniwa reviewed (commented):
Can we add a test with `LayoutTests/fast/dom/resources/leak-check.js` or other mechanism?

Ahmad-S792 commented on Source/WebCore/dom/BoundaryPoint.h:
```diff
@@ -79,4 +79,22 @@ inline BoundaryPoint makeBoundaryPointAfterNodeContents(Node& node)
+    , offset(offset)
+{
+}
+
```
nit: extra line.

NathanSolomon733 committed:
Document Leak occurs in LayoutTests/editing/mac/selection/word-thai.html
https://bugs.webkit.org/show_bug.cgi?id=280688
rdar://137058331

Reviewed by Ryan Reno.

The document leak occurs because the m_dragStartSelection member in
EventHandler, which is of type SimpleRange, holds Refs to Nodes.
This is a problem because the Document holds Refs to these nodes
because the document holds a strong reference to an Editor, which holds a
strong reference to a LocalFrame, which holds a strong reference to an
EventHandler.
This causes a reference cycle to occur between
the Document and the Nodes the EventHandler holds. Fix this reference
cycle by changing the type of m_dragStartSelection to a new type
that allows the member to hold a WeakReference to these nodes. Creating
this new type instead of modifying SimpleRange to hold WeakPtrs to nodes
prevents making extensive changes to areas of the codebase that can
still use a SimpleRange without causing a
reference cycle. Keep the change isolated to what code
relies on accessing m_dragStartSelection's Nodes to prevent adding
extensive amounts of null checking in other areas.

The test I ran to verify the fix was: run-webkit-tests --world-leaks LayoutTests/editing/mac/selection/word-thai.html
This test would previously fail when --world-leaks was passed in.

* Source/WebCore/dom/BoundaryPoint.h:
(WebCore::BoundaryPointWeak::BoundaryPointWeak):
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::SimpleRangeWeak::SimpleRangeWeak):
(WebCore::firstIntersectingNode):
(WebCore::nodePastLastIntersectingNode):
(WebCore::firstIntersectingNodeWithDeprecatedZeroOffsetStartQuirk):
(WebCore::commonInclusiveAncestor):
(WebCore::IntersectingNodeIterator::IntersectingNodeIterator):
* Source/WebCore/dom/SimpleRange.h:
(WebCore::SimpleRangeWeak::startContainer const):
(WebCore::SimpleRangeWeak::protectedStartContainer const):
(WebCore::SimpleRangeWeak::startOffset const):
(WebCore::SimpleRangeWeak::endContainer const):
(WebCore::SimpleRangeWeak::protectedEndContainer const):
(WebCore::SimpleRangeWeak::endOffset const):
(WebCore::SimpleRangeWeak::collapsed const):
(WebCore::SimpleRange::operator SimpleRangeWeak const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateSelectionForMouseDrag):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::treeOrderBoundaryPoints):

Canonical link: https://commits.webkit.org/285629@main
